internal/objects/service/service.go

```diff
@@ -379,3 +426,49 @@ func backendTCPNeeded(spec *operatorv1alpha1.ContourSpec) bool {
 		(spec.NetworkPublishing.Envoy.LoadBalancer.ProviderParameters.AWS == nil ||
 			spec.NetworkPublishing.Envoy.LoadBalancer.ProviderParameters.AWS.Type == operatorv1alpha1.AWSClassicLoadBalancer)
 }
+
+// allocationIDNeeded returns true if "service.beta.kubernetes.io/aws-load-balancer-eip-allocations"
+// annotation is needed based on the provided spec.
+func allocationIDNeeded(spec *operatorv1alpha1.ContourSpec) bool {
+	return spec.NetworkPublishing.Envoy.Type == operatorv1alpha1.LoadBalancerServicePublishingType &&
+		spec.NetworkPublishing.Envoy.LoadBalancer.ProviderParameters.Type == operatorv1alpha1.AWSLoadBalancerProvider &&
+		spec.NetworkPublishing.Envoy.LoadBalancer.ProviderParameters.AWS != nil &&
+		spec.NetworkPublishing.Envoy.LoadBalancer.ProviderParameters.AWS.Type == operatorv1alpha1.AWSNetworkLoadBalancer &&
+		len(spec.NetworkPublishing.Envoy.LoadBalancer.ProviderParameters.AWS.AllocationID) > 0
+}
```

Comment:
According to `awsLBAllocationIDAnnotation` you should also check for scope = "External".